### PR TITLE
completed validation module

### PR DESCRIPTION
**Pull request by Ptysiek**

### CODE REVIEW

**Comment by ziobron** [comment]
✅ CI Passed - Task: password-check-bonus
🏅 Granted 4 XP for @Ptysiek

**Comment by ziobron** [comment]
✅ CI Passed - Task: password-check
🏅 Granted 8 XP for @Ptysiek

**Comment by ziobron** [comment]
⏱ 3 XP extra for @Ptysiek for on-time delivery

File: homework/password-check/validation.cpp [hunk 1]
```diff
@@ -1,2 +1,57 @@
...
+        return "Password needs to have at least one uppercase letter";
+    case ErrorCode::PasswordsDoNotMatch:
+        return "Passwords do not match";
+    default:
+        return {};
+    }
+}
+
+bool doPasswordsMatch(const std::string& password, const std::string& passwordConfirm) {
+    return (password == passwordConfirm);
```
Comment (nauka-programowania-MA) [inline homework/password-check/validation.cpp:25]: `()` is redundant.

File: homework/password-check/validation.cpp [hunk 2]
```diff
@@ -1,2 +1,57 @@
...
+}
+
+template<class UnaryPredicate>
+bool notSingleCharacterIn(const std::string& str, UnaryPredicate unaryPredicate) {
+    return !std::any_of(str.begin(), str.end(), unaryPredicate);
```
Comment (nauka-programowania-MA) [inline homework/password-check/validation.cpp:30]: better us none_of

File: homework/password-check/validation.cpp [hunk 3]
```diff
@@ -1,2 +1,57 @@
...
+}
+
+ErrorCode checkPasswordRules(const std::string& password) {
+    if (password.length() < 9) {
+        return ErrorCode::PasswordNeedsAtLeastNineCharacters;
+    }
+    const auto isDigit = [](unsigned char character) { return std::isdigit(character); };
```
Comment (nauka-programowania-MA) [inline homework/password-check/validation.cpp:37]: One command one line

File: homework/password-check/validation.cpp [hunk 4]
```diff
@@ -1,2 +1,57 @@
...
+    }
+    const auto isSpecialCharacter = [&](unsigned char character) { return !(isDigit(character) || std::isalpha(character)); };
+    if (notSingleCharacterIn(password, isSpecialCharacter)) {
+        return ErrorCode::PasswordNeedsAtLeastOneSpecialCharacter;
+    }
+    const auto isUppercase = [](unsigned char character) { return std::isupper(character); };
+    if (notSingleCharacterIn(password, isUppercase)) {
+        return ErrorCode::PasswordNeedsAtLeastOneUppercaseLetter;
+    }
+    return ErrorCode::Ok;
```
Comment (nauka-programowania-MA) [inline homework/password-check/validation.cpp:49]: This is more compilacted then should be:

Just ` if (std::none_of(str.begin(), str.end(), isDigit)` It's clear and readable.

**Review by nauka-programowania-MA — APPROVED** [review]

**Pull request closed without merging**